docarray/array/storage/memory/backend.py

```diff
@@ -40,26 +54,28 @@ def _rebuild_id2offset(self) -> None:
             d.id: i for i, d in enumerate(self._data)
         }  # type: Dict[str, int]
 
+        self._needs_id2offset_rebuild = False
+
+    @needs_id2offset_rebuild
     def _init_storage(
         self, _docs: Optional['DocumentArraySourceType'] = None, copy: bool = False
     ):
         from ... import DocumentArray
 
         self._data = []
+        self._id_to_index = {}
         if _docs is None:
             return
         elif isinstance(
             _docs, (DocumentArray, Sequence, Generator, Iterator, itertools.chain)
         ):
             if copy:
                 self._data = [Document(d, copy=True) for d in _docs]
-                self._rebuild_id2offset()
             elif isinstance(_docs, DocumentArray):
                 self._data = _docs._data
                 self._id_to_index = _docs._id2offset
             else:
                 self._data = list(_docs)
```

Comment:
I'm not sure but does it make sense to copy the `self._needs_id2offset_rebuild` at this point ? (but only if it's also a `DocumentArrayInMemory`)
So maybe we can do:
```suggestion
            elif isinstance(_docs, DocumentArray):
                self._data = _docs._data
                self._id_to_index = _docs._id2offset
            else:
                self._data = list(_docs)
            
            if isinstance(_docs, DocumentArrayInMemory):
                self._needs_id2offset_rebuild = _docs._needs_id2offset_rebuild
```